src/mats_utils/geolocation/coordinates.py

```diff
@@ -178,4 +181,204 @@ def angles(ccditem):
+     at time t.
+    
+    
+    Parameters
+    ----------
+    s : float
+        length along the straight line
+    t : skyfield.timelib.Time
+        time
+    pos : array[float]
+        position in space where the line starts (~position of MATS). Array of 3 position coordinates in m in the ICRF reference frame
+    FOV : array[float]
+        angle of the line (direction of the line), array of 3 elements in the IFRC reference frame
+            
+    Returns
+    -------
+    elevation**2 : float
+        elevation squared of the point pos+s*FOV in m**2
+    """
+    newp = pos + s * FOV
+    newp = ICRF(Distance(m=newp).au, t=t, center=399)
+    return wgs84.subpoint(newp).elevation.m**2
+
+
+def findsurface(t, pos, FOV):
+    """
+    Function to get the distance between a point at position pos and the surface of the Geoid (wgs84 model),
+     at time t, along the line oriented along the FOV direction and starting at position pos
+    
+    
+    Parameters
+    ----------
+    t : skyfield.timelib.Time
+        time
+    pos : array[float]
+        position in space where the line starts (~position of MATS). Array of 3 position coordinates in m in the ICRF reference frame
+    FOV : array[float]
+        angle of the line (direction of the line), array of 3 elements in the IFRC reference frame
+            
+    Returns
+    -------
+    res : OptimizeResult object
+        res.x is the distance found in m   
+    """
+    res = minimize_scalar(funheight_square, args=(t, pos, FOV), bracket=(3e5, 8e5))
+    return res
+
+
+def NADIR_geolocation(ccditem,x_step=2,y_step=2):
+    """
+    Function to get the latitude, longitude and solar zenith angle map for each pixel of the image.
+    The values are calculated for some points and then interpolated for each pixel.
+    WARNING : no images are flipped
+    
+    Parameters
+    ----------
+    ccditem : CCDitem
+        measurement
+    x_step : int
+        step along the x-axis in the image between 2 sampled points used for interpolation. The default value is 2.
+    y_step : int
+        step along the y-axis in the image between 2 sampled points used for interpolation. The default value is 2.
+            
+    Returns
+    -------
+    lat_map : array[float]
+        map giving the latitude for each pixel in the image
+    lon_map : array[float]
+        map giving the longitude for each pixel in the image
+    sza_map : array[float]
+        map giving the solar zenith angle for each pixel in the image
+    """
+    im = ccditem['IMAGE']
+    x_deg_map, y_deg_map = deg_map(ccditem) # creating angle deviation map for each pixel (degress)
+    a,b = np.shape(im)
+
+    metoOHB  = R.from_matrix([[0,0,-1],[0,-1,0],[-1,0,0]])
+    ts=sfapi.load.timescale()
+    t=ts.from_datetime(ccditem['EXPDate'].replace(tzinfo=sfapi.utc)) # exposure time  
+    q=ccditem.afsAttitudeState
+    quat=R.from_quat(np.roll(q,-1)) # quaternion of MATS attitude (for the LIMB imager) 
```

Comment:
the quaternion is not for the limb imager but for the satellite frame.

Reply:
done